Lib/mimetypes.py

```diff
@@ -562,6 +590,153 @@ def _default_mime_types():
+    for idx, hb in enumerate(h):
+        if not hb in b'\t\n\x0c\r ':
+            first_non_ws = idx
+            break
+
+    detects= [_match_html_types, _match_exact_sig_types,
+              _match_mask_sig_types, _match_mp4_type, _match_text_type]
+    for detect in detects:
+        tp, charset = detect(h, first_non_ws)
+        if tp:
+            return (tp, charset)
+    return (None, None)
+
+
+def _match_html_types(h, first_non_ws):
+    sigs = [b'<!DOCTYPE HTML', b'<HTML', b'<HEAD',
+            b'<SCRIPT', b'<IFRAME', b'<H1', b'<DIV',
+            b'<FONT', b'<TABLE', b'<A', b'<STYLE',
+            b'<TITLE', b'<B', b'<BODY', b'<BR', b'<P',
+            b'<!--']
+    h = h[first_non_ws:]
+
+    for s in sigs:
+        tp, charset = _match_html_sig(h, s)
+        if tp:
+            return (tp, charset)
+
+    return (None, None)
+
+
+def _match_html_sig(h, sig):
+    tp, charset = None, None
+    if len(h) < len(sig) + 1:
+        return (tp, charset)
+
+    for hc, sc in zip(h, sig):
+        if 65 <= sc <= 90:
+            hc&=0xDF
+        if hc != sc:
+            return (tp, charset)
+
+    # should be a tag-terminating byte (0xTT)
+    # https://mimesniff.spec.whatwg.org/#terminology
+    if not h[len(sig)] in b' >':
+        return (tp, charset)
+    tp, charset = 'text/html', 'utf-8'
+    return (tp, charset)
+
+
+def _match_exact_sig_types(h, first_non_ws):
+    sigs = [
+        # (pattern, mimetype)
+        (b'\x00\x00\x01\x00', 'image/x-icon'),
+        (b'\x00\x00\x02\x00', 'image/x-icon'),
+        (b'%PDF-', 'application/pdf'),
+        (b'%!PS-Adobe-', 'application/postscript'),
+        (b'BM', 'image/bmp'),
+        (b'GIF87a', 'image/gif'),
+        (b'GIF89a', 'image/gif'),
+        (b'\x89PNG\x0D\x0A\x1A\x0A', 'image/png'),
+        (b'\xFF\xD8\xFF', 'image/jpeg'),
+        (b'\x00\x01\x00\x00', 'font/ttf'),
+        (b'OTTO', 'font/otf'),
+        (b'ttcf', 'font/collection'),
+        (b'wOFF', 'font/woff'),
+        (b'wOF2', 'font/woff2'),
+        (b'\x1F\x8B\x08', 'application/x-gzip'),
+        (b'PK\x03\x04', 'application/zip'),
+        (b'Rar!\x1A\x07\x00', 'application/x-rar-compressed'),
+        (b'Rar!\x1A\x07\x01\x00', 'application/x-rar-compressed'),
+    ]
+
+    h = h[first_non_ws:]
+    for sig, mime in sigs:
+        if h.startswith(sig):
+            return (mime, None)
+
+    return (None, None)
+
+
+def _match_mask_sig_types(h, first_non_ws):
+    sigs = [
+        # (mask, pattern, skip_white_space, mimetype)
+        (b'\xFF\xFF\xFF\xFF\xFF', b'<?xml', True, ('text/xml','utf-8')),
+        (b'\xFF\xFF\x00\x00', b'\xFE\xFF\x00\x00', False, ('text/plain', 'utf-16be')),
+        (b'\xFF\xFF\x00\x00', b'\xFF\xFE\x00\x00', False, ('text/plain', 'utf-16le')),
+        (b'\xFF\xFF\xFF\x00', b'\xEF\xBB\xBF\x00', False, ('text/plain', 'utf-8')),
+        (b'\xFF\xFF\xFF\xFF\x00\x00\x00\x00\xFF\xFF\xFF\xFF\xFF\xFF', b'RIFF\x00\x00\x00\x00WEBPVP', False, ('image/webp', None)),
```

Comment:
Nitpick: It would be nice reformat these longer lines. For example:

```py
sigs = [
    ...
    (
        b'\xFF\xFF\xFF\xFF\x00\x00\x00\x00\xFF\xFF\xFF\xFF\xFF\xFF',
        b'RIFF\x00\x00\x00\x00WEBPVP',
        False,
        ('image/webp', None),
    ),
    ...
]
```